build/v1/types.go

```diff
@@ -1359,3 +1368,64 @@ type SecretSpec struct {
 }
+
+// BuildVolume describes a volume that is made available to build pods,
+// such that it can be mounted into buildah's runtime environment.
+// Only a subset of Kubernetes Volume sources are supported.
+type BuildVolume struct {
+	// name is a unique identifier for this BuildVolume.
+	// It must conform to the Kubernetes DNS label standard and be unique within the pod.
+	// It cannot collide with volumes that are always added by the build controller.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// source represents the location and type of the mounted volume.
+	Source BuildVolumeSource `json:"source" protobuf:"bytes,2,opt,name=source"`
+
+	// mounts represents the location and type of the mounted volume.
+	Mounts []BuildVolumeMount `json:"mounts" protobuf:"bytes,3,opt,name=mounts"`
+}
+
+// BuildVolumeSourceType represents a build volume source type
+type BuildVolumeSourceType string
+
+const (
+	// BuildVolumeSourceTypeSecret is the Secret build source volume type
+	BuildVolumeSourceTypeSecret BuildVolumeSourceType = "Secret"
+
+	// BuildVolumeSourceTypeConfigmap is the ConfigMap build source volume type
+	BuildVolumeSourceTypeConfigMap BuildVolumeSourceType = "ConfigMap"
+)
+
+// BuildVolumeSource represents the source of a volume to mount
+// Only one of its supported types may be specified at any given time.
+type BuildVolumeSource struct {
+
+	// type is the BuildVolumeSourceType for the volume source.
+	// Type must match the populated volume source, and if not specified will be inferred on admission.
+	// +k8s:conversion-gen=false
+	// +optional
+	Type BuildVolumeSourceType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=BuildVolumeSourceType"`
+
+	// secret represents a Secret that should populate this volume.
+	// More info: https://kubernetes.io/docs/concepts/storage/volumes#secret
+	// +optional
+	Secret *corev1.SecretVolumeSource `json:"secret,omitempty" protobuf:"bytes,2,opt,name=secret"`
+
+	// configMap represents a ConfigMap that should populate this volume
+	// +optional
+	ConfigMap *corev1.ConfigMapVolumeSource `json:"configMap,omitempty" protobuf:"bytes,3,opt,name=configMap"`
+}
+
+// BuildVolumeMount describes the mounting of a Volume within buildah's runtime environment.
+type BuildVolumeMount struct {
+	// destinationPath is the path within the build runtime environment at which the volume should be mounted.
+	// Must not contain ':' and can not collide with a destination path generated by the builder process.
+	DestinationPath string `json:"destinationPath" protobuf:"bytes,1,opt,name=destinationPath"`
+
+	// readOnly denotes whether the volume should be mounted as readonly or not
+	// The volume is mounted read-only if true, read-write otherwise (false or unspecified).
+	// Defaults to false.
+	// +optional
+	ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,2,opt,name=readOnly"`
```

Comment:
This implies that when we set up the volume mount in the build pod containers, we either:

1. Always mount as read-write.
2. Be smart about setting read-only. For example, set the build container volume mount as read-only if and only if all buildah volume mounts are read only.

@nalind can we run into issues if the same directory is mounted into buildah in two different locations with different read/write settings?

@jsafrane must we always mount Secrets and ConfigMaps into containers as read-only? Is mounting a Secret or ConfigMap as a writeable directory impossible, or merely a bad idea?

Reply:
@bparees So, during our discussion I thought that we clarified that the volumeMounts don't need to be setup in the container? Only the buildah mounts?

Reply:
we clarified that you have total freedom to decide how to get the content into the build pod/container (using adam's terminology).  It will probably be *a* volume/volumeount, but it doesn't have to be the volume/volumemount as specified by the user (in terms of the volume name and the volumemount path).

Reply:
as regards read/readwrite...  for secrets/configmaps it really doesn't matter (nothing you do is going to change the actual secret/configmap resource).  If/when this is extended to cover persistent volumes for caching build data, how you mount the volume into the build pod/container will matter, however.

Reply:
> must we always mount Secrets and ConfigMaps into containers as read-only?

Yes. See here: https://github.com/kubernetes/kubernetes/issues/62099 (+ linked PR)

Reply:
@bparees @coreydaley thoughts on dropping this field, and later introducing a field like `Writeable` for read-write volumes?

The current implementation has us do the following:

1. Mount the volume into the builder container in a fixed directory - ex `/var/run/build/volumes/xxxx`
2. Tell buildah to use the directory above as a transient bind mount, passing the `ro` or `rw` option as specified by the user.

If a Secret or ConfigMap must always be mounted as a read only filesystem, then having the user bind mount said read-only filesystem in buildah as a writeable directory will lead to inevitable errors. To avoid these errors, we would need to copy the contents to a writeable directory, which feels like overkill.

Reply:
Dropping the field for now makes sense to me, since it is not needed.

Reply:
@adambkaplan updated in current push, ptal